### PR TITLE
[Impeller] Implement morphology image filter

### PR DESCRIPTION
**Pull request by ColdPaleLight**
fix https://github.com/flutter/flutter/issues/109731
fix https://github.com/flutter/flutter/issues/109730

https://user-images.githubusercontent.com/31977171/185794393-5e8511a6-b708-42e8-a746-c850886da251.mp4

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.

### CODE REVIEW

File: impeller/entity/contents/filters/morphology_filter_contents.cc [hunk 1]
```diff
@@ -0,0 +1,195 @@
...
+
+  auto transformed_radius = entity.GetTransformation().TransformDirection(
+      direction_ * radius_.radius);
+
+  auto transformed_radius_length = transformed_radius.GetLength();
+
+  // A matrix that rotates the snapshot space such that the direction is
+  // +X.
+  auto texture_rotate =
+      Matrix::MakeRotationZ(transformed_radius.Normalize().AngleTo({1, 0}));
```
Comment (bdero) [inline impeller/entity/contents/filters/morphology_filter_contents.cc:70]: Flipping the texture to face +X direction was only needed in the blur filter for downsampling optimizations and this complexity can be avoided for this filter. I'd recommend taking a look at the blur filter's code before the recent downsampling optimization. The UV mapping situation was a bit simpler and `Snapshot` has a utility that can be used to compute the UVs here to great effect!

**Review by bdero — COMMENTED** [review]
Awesome, thanks for taking this on! I added a comment about reducing complexity of the UV mapping, but this is overall looking great.

File: impeller/entity/contents/filters/morphology_filter_contents.cc (hunk 1 above)
Comment (ColdPaleLight) [reply impeller/entity/contents/filters/morphology_filter_contents.cc]: Done.

**Review by bdero — APPROVED** [review]
LGTM, thanks! We've been landing a lot of patches around filters and snapshots lately, so consider giving it one more rebase before merging for safety.

File: impeller/entity/shaders/morphology_filter.frag [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+frag_info;
+
+in vec2 v_texture_coords;
+out vec4 frag_color;
+
+void main() {
+  vec4 result = frag_info.morph_type == kMorphTypeDilate ? vec4(0) : vec4(1);
+  vec2 uv_offset = frag_info.direction / frag_info.texture_size;
+  for (float i = -frag_info.radius; i <= frag_info.radius; i++) {
+    vec2 texture_coords = v_texture_coords + uv_offset * i;
```
Comment (fzyzcjy) [inline impeller/entity/shaders/morphology_filter.frag:31]: Not an expert for shaders, but can we really do this "i++" for *float* radius - will we miss 0.0~1.0 pixels in the end?

**Comment by fzyzcjy** [comment]
(Btw, good job!)

File: impeller/entity/shaders/morphology_filter.frag (hunk 2 above)
Comment (ColdPaleLight) [reply impeller/entity/shaders/morphology_filter.frag:31]: Since the radius is converted to an integer by the std::round method, i++ can be used here :)

Comment (fzyzcjy) [reply impeller/entity/shaders/morphology_filter.frag:31]: Thanks! I see. The fact that it is a "float" makes me confused at the first glance though :)

**Pull request merged**